blog/_src/posts/2017-02-27-metafunction-apply.md

```diff
@@ -0,0 +1,214 @@
+ for building a list, Redex doesn't report an error.
+
+We can fix this code with the new
+[`mf-apply`](https://www.cs.utah.edu/plt/snapshots/current/doc/redex/The_Redex_Reference.html#%28form._%28%28lib._redex%2Freduction-semantics..rkt%29._mf-apply%29%29)
+keyword (available on [GitHub](https://github.com/racket/racket) or in a
+[snapshot build](https://www.cs.utah.edu/plt/snapshots/)):
+
+```racket
+   ....
+   (where ((x_1 τ_1) ...) (mf-apply closure-env C))
+   ....
+```
+
+Running `raco make` now gives a compile-time error.
+
+```
+  term: expected a previously defined metafunction
+    at: closure-env
+    in: (mf-apply closure-env C)
+```
+
+
+### But I still need to type `mf-apply` correctly!
+
+Leif Andersen says:
+
+> I should point out that this has the issue of you still need to type
+> `mf-apply` correctly. ;)
+
+That is, if you accidentally write:
+
+```racket
+   ....
+   (where ((x_1 τ_1) ...) (mf-applu closure-env C))
+   ....
+```
+
+Then the code compiles, thinking you intend to match a list of three elements
+ against the pattern.
+
+Never fear, there are at least two solutions.
+
+
+#### Solution 1: rename `mf-apply`
+
+A simple fix is to rename the `mf-apply` keyword to something shorter:
+
+```racket
+#lang racket
+(require redex
+         (rename-in redex
+           [mf-apply MF]))
+```
+
+
+#### Solution 2: the `mf-apply` lang extension
+
+A fancier solution is to install the `mf-apply` meta-language.
+
+```
+  $ raco pkg install mf-apply
+```
+
+This language updates the [_readtable_](http://docs.racket-lang.org/reference/readtables.html#%28tech._readtable%29)
+ to interpret S-expressions that begin with `#{`:
+
+```racket
+#lang mf-apply racket
+(require redex)
+
+(term #{f x ...})
+```
+
+as a metafunction application:
+
+```racket
+#lang mf-apply racket
+(require redex)
+
+(term (mf-apply f x ...))
+```
+
+Source code is on GitHub:
+
+- [https://github.com/bennn/mf-apply](https://github.com/bennn/mf-apply)
+
+(It's the simplest lang-extension I know of)
+
+
+## What is PLT Redex?
+
+PLT Redex is a library for semantics engineering.
+One of my favorite Redex features is it implements capture-avoiding substitution
+ and α-equivalence for any language with a `#:binding-forms` specification
+ (such as STLC, above).
+
+You can read more:
+
+- in the "Amb" tutorial: [http://docs.racket-lang.org/redex/tutorial.html](http://docs.racket-lang.org/redex/tutorial.html)
+- in the "Long Tutorial": [http://docs.racket-lang.org/redex/redex2015.html](http://docs.racket-lang.org/redex/redex2015.html)
+- in the Redex reference manual: [http://docs.racket-lang.org/redex/The_Redex_Reference.html](http://docs.racket-lang.org/redex/The_Redex_Reference.html)
+- on the PLT Redex website: [https://redex.racket-lang.org/](https://redex.racket-lang.org/)
+- on GitHub: [https://github.com/racket/redex](https://github.com/racket/redex)
+
+And if you act now, you can become a _Redexionsita_ between July 10 and July 14
```

Comment:
Talked with Max, Leif, and Ming-Ho about this as lunch. Some more ideas:
- redexan
- redexhibistionist (my favorite)
- redexist